File: Visual Studio Projects/WalkTreeExample/WalkTreeExample/WalkTree.cs
/* --------------------------------------------------------------------------------
 *  <copyright file="WalkTree.cs" company="Hybrid Software Helix Ltd">
 *    Copyright (c) 2025 Hybrid Software Helix Ltd. All rights reserved.
 *  </copyright>
 *  <summary>
 *    This example is provided on an "as is" basis and without warranty of any kind.
 *    Hybrid Software Helix Ltd. does not warrant or make any representations
 *    regarding the use or results of use of this example.
 *  </summary>
 * ---------------------------------------------------------------------------------
 */

using JawsMako;

namespace WalkTreeExample;

internal class WalkTree
{
    static int Main()
    {
        try
        {
            var testFilepath = @"..\..\..\..\TestFiles\";

            var mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);

            // Input
            var pdfInput = IPDFInput.create(mako);
            using var assembly = pdfInput.open(testFilepath + "Cheshire Cat.pdf");
            using var inFixedPage = assembly.getDocument().getPage()
                .getContent();

            //void VisitNodeHandler(IDOMNode node) => DumpNodeInfo(node);
            //WalkTreeCallback callback = new NodeTreeCallbackDelegate(VisitNodeHandler);

            WalkTreeCallback callback = new NodeTreeCallback();

            inFixedPage.walkTree(callback.getCallbackFunc(), callback.getPriv(), false, true);

        }
        catch (MakoException e)
        {
            Console.WriteLine($"Exception thrown: {e.m_errorCode}: {e.m_msg}");
        }
        catch (Exception e)
        {
            Console.WriteLine($"Exception thrown: {e}");
        }

        return 0;
    }

    private class NodeTreeCallback : WalkTreeCallback
    {
        public override bool visitNode(IDOMNode node)
        {
            DumpNodeInfo(node);
            return true;
        }
    }

    // ReSharper disable once UnusedType.Local
    private class NodeTreeCallbackDelegate(NodeTreeCallbackDelegate.visitNodeDelegate visitNodeHandler) : WalkTreeCallback
    {
        public delegate void visitNodeDelegate(IDOMNode node);

        public override bool visitNode(IDOMNode node)
        {
            if (visitNodeHandler == null)
                return false;
            visitNodeHandler(node);
            return true;
        }
    }

    static void DumpNodeInfo(IDOMNode node)
    {
        IDOMNode p = node;
        while (p.getParentNode() is not null)
        {
            Console.Write("  ");
            p = p.getParentNode();
        }

        switch (node.getNodeType())
        {
            case eDOMNodeType.eDOMFixedPageNode:
                Console.Write("eDOMFixedPageNode");
                break;
            case eDOMNodeType.eDOMGroupNode:
                Console.Write("eDOMGroupNode");
                break;
            case eDOMNodeType.eDOMCharPathGroupNode:
                Console.Write("eDOMCharPathGroupNode");
                break;
            case eDOMNodeType.eDOMTransparencyGroupNode:
                Console.Write("eDOMTransparencyGroupNode");
                break;
            case eDOMNodeType.eDOMGlyphsNode:
                Console.Write("eDOMGlyphsNode");
                break;
            case eDOMNodeType.eDOMPathNode:
                Console.Write("eDOMPathNode");
                break;
            case eDOMNodeType.eDOMFormNode:
                Console.Write("eDOMFormNode");
                break;
            case eDOMNodeType.eDOMFormInstanceNode:
                Console.Write("eDOMFormInstanceNode");
                break;
            case eDOMNodeType.eDOMContentRootNode:
                Console.Write("eDOMContentRootNode");
                break;
            case eDOMNodeType.eDOMDocumentSequenceNode:
                Console.Write("eDOMDocumentSequenceNode");
                break;
            case eDOMNodeType.eDOMDocumentNode:
                Console.Write("eDOMDocumentNode");
                break;
            case eDOMNodeType.eDOMFixedDocumentNode:
                Console.Write("eDOMFixedDocumentNode");
                break;
            case eDOMNodeType.eDOMPageNode:
                Console.Write("eDOMPageNode");
                break;
            case eDOMNodeType.eDOMCanvasNode:
                Console.Write("eDOMCanvasNode");
                break;
            case eDOMNodeType.eDOMGlyphNode:
                Console.Write("eDOMGlyphNode");
                break;
            case eDOMNodeType.eDOMRefNode:
                Console.Write("eDOMRefNode");
                break;
            case eDOMNodeType.eDOMJobTkContentNode:
                Console.Write("eDOMJobTkContentNode");
                break;
            case eDOMNodeType.eDOMJobTkNodeNode:
                Console.Write("eDOMJobTkNodeNode");
                break;
            case eDOMNodeType.eDOMJobTkValueNode:
                Console.Write("eDOMJobTkValueNode");
                break;
            case eDOMNodeType.eDOMJobTkGenericNodeNode:
                Console.Write("eDOMJobTkGenericNodeNode");
                break;
            case eDOMNodeType.eDOMJobTkGenericCharacterDataNode:
                Console.Write("eDOMJobTkGenericCharacterDataNode");
                break;
            case eDOMNodeType.eDOMVisualRootNode:
                Console.Write("eDOMVisualRootNode");
                break;
            case eDOMNodeType.eDOMOutlineNode:
                Console.Write("eDOMOutlineNode");
                break;
            case eDOMNodeType.eDOMOutlineEntryNode:
                Console.Write("eDOMOutlineEntryNode");
                break;
            case eDOMNodeType.eDOMAnnotationAppearanceNode:
                Console.Write("eDOMAnnotationAppearanceNode");
                break;
            case eDOMNodeType.eDOMFragmentNode:
                Console.Write("eDOMFragmentNode");
                break;
            case eDOMNodeType.eDOMTileNode:
                Console.Write("eDOMTileNode");
                break;
            case eDOMNodeType.eDOMNodeTypeCnt:
                Console.Write("eDOMNodeTypeCnt");
                break;
        }
        Console.WriteLine();
    }
}

File: Java Projects/WalkTreeExample/src/WalkTree.java
/* --------------------------------------------------------------------------------
 *  <copyright file="WalkTree.java" company="Hybrid Software Helix Ltd">
 *    Copyright (c) 2025 Hybrid Software Helix Ltd. All rights reserved.
 *  </copyright>
 *  <summary>
 *    This example is provided on an "as is" basis and without warranty of any kind.
 *    Hybrid Software Helix Ltd. does not warrant or make any representations
 *    regarding the use or results of use of this example.
 *  </summary>
 * ---------------------------------------------------------------------------------
 */

import com.globalgraphics.JawsMako.jawsmakoIF.*;
import com.globalgraphics.JawsMako.jawsmakoIF.jawsmakoIF.*;

public class WalkTree
{
    public static void main(String[] args)
    {
        try
        {
            String testFilepath = "TestFiles/";

            // Instantiate Mako
            IJawsMako mako = IJawsMako.create();
            IJawsMako.enableAllFeatures(mako);

            // Input
            IPDFInput pdfInput = IPDFInput.create(mako);
            IDocumentAssembly assembly = pdfInput.open(testFilepath + "Cheshire Cat.pdf");

            IDOMFixedPage inFixedPage = assembly.getDocument().getPage(0).getContent();

            // Define the callback and walk the tree
            WalkTreeCallback callback = new NodeTreeCallback();
            inFixedPage.walkTree(callback.getCallbackFunc(), callback.getPriv(), false, true);
        }
        catch (Exception e)
        {
            System.out.printf("Exception thrown: %s%n", e.toString());
        }
    }

    private static class NodeTreeCallback extends WalkTreeCallback
    {
        @Override
        public boolean visitNode(IDOMNode node)
        {
            dumpNodeInfo(node);
            return true;
        }
    }

    private static void dumpNodeInfo(IDOMNode node)
    {
        IDOMNode p = node;
        while (p.getParentNode() != null)
        {
            System.out.print("  ");
            p = p.getParentNode();
        }

        switch (node.getNodeType())
        {
            case eDOMFixedPageNode: System.out.print("eDOMFixedPageNode"); break;
            case eDOMGroupNode: System.out.print("eDOMGroupNode"); break;
            case eDOMCharPathGroupNode: System.out.print("eDOMCharPathGroupNode"); break;
            case eDOMTransparencyGroupNode: System.out.print("eDOMTransparencyGroupNode"); break;
            case eDOMGlyphsNode: System.out.print("eDOMGlyphsNode"); break;
            case eDOMPathNode: System.out.print("eDOMPathNode"); break;
            case eDOMFormNode: System.out.print("eDOMFormNode"); break;
            case eDOMFormInstanceNode: System.out.print("eDOMFormInstanceNode"); break;
            case eDOMContentRootNode: System.out.print("eDOMContentRootNode"); break;
            case eDOMDocumentSequenceNode: System.out.print("eDOMDocumentSequenceNode"); break;
            case eDOMDocumentNode: System.out.print("eDOMDocumentNode"); break;
            case eDOMFixedDocumentNode: System.out.print("eDOMFixedDocumentNode"); break;
            case eDOMPageNode: System.out.print("eDOMPageNode"); break;
            case eDOMCanvasNode: System.out.print("eDOMCanvasNode"); break;
            case eDOMGlyphNode: System.out.print("eDOMGlyphNode"); break;
            case eDOMRefNode: System.out.print("eDOMRefNode"); break;
            case eDOMJobTkContentNode: System.out.print("eDOMJobTkContentNode"); break;
            case eDOMJobTkNodeNode: System.out.print("eDOMJobTkNodeNode"); break;
            case eDOMJobTkValueNode: System.out.print("eDOMJobTkValueNode"); break;
            case eDOMJobTkGenericNodeNode: System.out.print("eDOMJobTkGenericNodeNode"); break;
            case eDOMJobTkGenericCharacterDataNode: System.out.print("eDOMJobTkGenericCharacterDataNode"); break;
            case eDOMVisualRootNode: System.out.print("eDOMVisualRootNode"); break;
            case eDOMOutlineNode: System.out.print("eDOMOutlineNode"); break;
            case eDOMOutlineEntryNode: System.out.print("eDOMOutlineEntryNode"); break;
            case eDOMAnnotationAppearanceNode: System.out.print("eDOMAnnotationAppearanceNode"); break;
            case eDOMFragmentNode: System.out.print("eDOMFragmentNode"); break;
            case eDOMTileNode: System.out.print("eDOMTileNode"); break;
            case eDOMNodeTypeCnt: System.out.print("eDOMNodeTypeCnt"); break;
        }

        System.out.println();
    }
}
